AssertionError in tpd.newArray on new Array of opaque type

In the companion object of an opaque type `T`, creating an array using `new Array[T]` causes an assertion error. Seems to work outside the companion.
Code:
```scala
object A {
  opaque type T = Int
  object T{
    val a = new Array[T](3)
  }
}
```
Output:
```
exception occurred while compiling opaque-array.scala
java.lang.AssertionError: assertion failed while compiling opaque-array.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.Predef$.assert(Predef.scala:208)
        at dotty.tools.dotc.ast.tpd$.newArray(tpd.scala:410)
        at dotty.tools.dotc.transform.ArrayConstructors.expand$1(ArrayConstructors.scala:29)
        at dotty.tools.dotc.transform.ArrayConstructors.transformApply(ArrayConstructors.scala:33)
        at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:611)
        at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:612)
        at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:612)
        at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:612)
        at dotty.tools.dotc.transform.MegaPhase.goApply(MegaPhase.scala:612)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:272)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
        at dotty.tools.dotc.transform.MegaPhase.mapValDef$1(MegaPhase.scala:234)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:238)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
        at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:404)
        at dotty.tools.dotc.transform.MegaPhase.$anonfun$transformStats$2(MegaPhase.scala:409)
        at scala.collection.immutable.List.mapConserve(List.scala:179)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:339)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
        at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:404)
        at dotty.tools.dotc.transform.MegaPhase.$anonfun$transformStats$2(MegaPhase.scala:409)
        at scala.collection.immutable.List.mapConserve(List.scala:179)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:339)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:394)
        at dotty.tools.dotc.transform.MegaPhase.transformStat$1(MegaPhase.scala:404)
        at dotty.tools.dotc.transform.MegaPhase.$anonfun$transformStats$2(MegaPhase.scala:409)
        at scala.collection.immutable.List.mapConserve(List.scala:179)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:409)
        at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:356)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:359)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:396)
        at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:415)
        at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:427)
        at dotty.tools.dotc.core.Phases$Phase.$anonfun$runOn$1(Phases.scala:316)
        at scala.collection.immutable.List.map(List.scala:286)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:314)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:158)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:49)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:155)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:153)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:153)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:178)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:102)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:133)
        at dotty.tools.dotc.Run.compileSources(Run.scala:120)
        at dotty.tools.dotc.Run.compile(Run.scala:104)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:33)
        at dotty.tools.dotc.Driver.process(Driver.scala:170)
        at dotty.tools.dotc.Driver.process(Driver.scala:139)
        at dotty.tools.dotc.Driver.process(Driver.scala:151)
        at dotty.tools.dotc.Driver.main(Driver.scala:178)
        at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -407,7 +407,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       ref(defn.DottyArraysModule).select(defn.newArrayMethod).withSpan(span)
 
     if (!ctx.erasedTypes) {
-      assert(!TypeErasure.isGeneric(elemTpe)) //needs to be done during typer. See Applications.convertNewGenericArray
+      assert(!TypeErasure.isGeneric(elemTpe), elemTpe) //needs to be done during typer. See Applications.convertNewGenericArray
       newArr.appliedToTypeTrees(TypeTree(returnTpe) :: Nil).appliedToArgs(clsOf(elemTpe) :: clsOf(returnTpe) :: dims :: Nil).withSpan(span)
     } else  // after erasure
       newArr.appliedToArgs(clsOf(elemTpe) :: clsOf(returnTpe) :: dims :: Nil).withSpan(span)

--- a/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
+++ b/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
@@ -61,13 +61,13 @@ object DenotTransformers {
   trait SymTransformer extends DenotTransformer {
 
   	/** Tramsform the info of a denotation that is not a Symdenotation */
-    def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = tp
+    def transformNonSymInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = tp
 
     def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation
 
     def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = ref match {
       case ref: SymDenotation => transformSym(ref)
-      case _ => ref.derivedSingleDenotation(ref.symbol, transformInfo(ref.info, ref.symbol))
+      case _ => ref.derivedSingleDenotation(ref.symbol, transformNonSymInfo(ref.info, ref.symbol))
     }
   }
 

--- a/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
+++ b/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
@@ -57,17 +57,18 @@ object DenotTransformers {
     protected def mayChange(sym: Symbol)(implicit ctx: Context): Boolean = true
   }
 
-  /** A transformer that only transforms SymDenotations */
+  /** A transformer that only transforms SymDenotations.
+   *  Note: Infos of non-sym denotations are left as is. So the transformer should
+   *  be used before erasure only if this is not a problem. After erasure, all
+   *  denotations are SymDenotations, so SymTransformers can be used freely.
+   */
   trait SymTransformer extends DenotTransformer {
-
-  	/** Tramsform the info of a denotation that is not a Symdenotation */
-    def transformNonSymInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = tp
 
     def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation
 
     def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = ref match {
       case ref: SymDenotation => transformSym(ref)
-      case _ => ref.derivedSingleDenotation(ref.symbol, transformNonSymInfo(ref.info, ref.symbol))
+      case _ => ref
     }
   }
 

--- a/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
+++ b/compiler/src/dotty/tools/dotc/core/DenotTransformers.scala
@@ -60,11 +60,14 @@ object DenotTransformers {
   /** A transformer that only transforms SymDenotations */
   trait SymTransformer extends DenotTransformer {
 
+  	/** Tramsform the info of a denotation that is not a Symdenotation */
+    def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = tp
+
     def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation
 
     def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = ref match {
       case ref: SymDenotation => transformSym(ref)
-      case _ => ref
+      case _ => ref.derivedSingleDenotation(ref.symbol, transformInfo(ref.info, ref.symbol))
     }
   }
 

--- a/compiler/src/dotty/tools/dotc/core/SymDenotations.scala
+++ b/compiler/src/dotty/tools/dotc/core/SymDenotations.scala
@@ -19,6 +19,7 @@ import reporting.diagnostic.Message
 import reporting.diagnostic.messages.BadSymbolicReference
 import reporting.trace
 import collection.mutable
+import transform.TypeUtils._
 
 import scala.annotation.internal.sharable
 
@@ -1082,16 +1083,10 @@ object SymDenotations {
      *  containing object.
      */
     def opaqueAlias(implicit ctx: Context): Type = {
-      if (isOpaqueHelper) {
+      if (isOpaqueHelper)
         owner.asClass.classInfo.selfType match {
-          case RefinedType(_, _, TypeBounds(lo, _)) =>
-            def extractAlias(tp: Type): Type = tp match {
-              case OrType(alias, _) => alias
-              case tp: HKTypeLambda => tp.derivedLambdaType(resType = extractAlias(tp.resType))
-            }
-            extractAlias(lo)
+          case RefinedType(_, _, bounds) => bounds.extractOpaqueAlias
         }
-      }
       else NoType
     }
 

--- a/compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala
+++ b/compiler/src/dotty/tools/dotc/printing/RefinedPrinter.scala
@@ -254,7 +254,13 @@ class RefinedPrinter(_ctx: Context) extends PlainPrinter(_ctx) {
   protected def typeApplyText[T >: Untyped](tree: TypeApply[T]): Text = {
     val isQuote = tree.fun.hasType && tree.fun.symbol == defn.InternalQuoted_typeQuote
     val (open, close) = if (isQuote) (keywordStr("'["), keywordStr("]")) else ("[", "]")
-    toTextLocal(tree.fun).provided(!isQuote) ~ open ~ toTextGlobal(tree.args, ", ") ~ close
+    val funText = toTextLocal(tree.fun).provided(!isQuote)
+    tree.fun match {
+      case Select(New(tpt), nme.CONSTRUCTOR) if tpt.typeOpt.dealias.isInstanceOf[AppliedType] =>
+        funText  // type was already printed by toText(new)
+      case _ =>
+        funText ~ open ~ toTextGlobal(tree.args, ", ") ~ close
+    }
   }
 
   protected def toTextCore[T >: Untyped](tree: Tree[T]): Text = {

--- a/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
@@ -2,25 +2,15 @@ package dotty.tools.dotc
 package transform
 
 import core._
-import Names._
 import dotty.tools.dotc.transform.MegaPhase._
-import ast.Trees._
-import ast.untpd
 import Flags._
 import Types._
-import Constants.Constant
 import Contexts.Context
 import Symbols._
 import Decorators._
-import Annotations._
-import Annotations.ConcreteAnnotation
-import Denotations.SingleDenotation
 import SymDenotations.SymDenotation
-import scala.collection.mutable
 import DenotTransformers._
-import NameOps._
-import NameKinds.OuterSelectName
-import StdNames._
+import TypeUtils._
 
 object ElimOpaque {
   val name: String = "elimOpaque"
@@ -36,6 +26,9 @@ class ElimOpaque extends MiniPhase with SymTransformer {
 
   // base types of opaque aliases change
   override def changesBaseTypes = true
+
+  override def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type =
+    if (sym.isOpaqueHelper) TypeAlias(tp.extractOpaqueAlias) else tp
 
   def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation =
     if (sym.isOpaqueHelper) {

--- a/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
@@ -28,15 +28,21 @@ class ElimOpaque extends MiniPhase with DenotTransformer {
   // base types of opaque aliases change
   override def changesBaseTypes = true
 
-  def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation =
-    if (ref.symbol.isOpaqueHelper)
-      ref match {
-        case sym: SymDenotation =>
-          sym.copySymDenotation(
-            info = TypeAlias(sym.opaqueAlias),
-            initFlags = sym.flags &~ (Opaque | Deferred))
-        case _ =>
-          ref.derivedSingleDenotation(ref.symbol, TypeAlias(ref.info.extractOpaqueAlias))
+  def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = {
+    val sym = ref.symbol
+    ref match {
+      case ref: SymDenotation if sym.isOpaqueHelper =>
+        ref.copySymDenotation(
+          info = TypeAlias(ref.opaqueAlias),
+          initFlags = ref.flags &~ (Opaque | Deferred))
+      case ref: SymDenotation if sym.isOpaqueCompanion =>
+        val ref1 = ref.copySymDenotation(initFlags = ref.flags &~ Opaque)
+        ref1.registeredCompanion = NoSymbol
+        ref1
+      case _ if sym.isOpaqueHelper =>
+        ref.derivedSingleDenotation(sym, TypeAlias(ref.info.extractOpaqueAlias))
+      case _ =>
+        ref
     }
-    else ref
+  }
 }

--- a/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
@@ -27,7 +27,7 @@ class ElimOpaque extends MiniPhase with SymTransformer {
   // base types of opaque aliases change
   override def changesBaseTypes = true
 
-  override def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type =
+  override def transformNonSymInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type =
     if (sym.isOpaqueHelper) TypeAlias(tp.extractOpaqueAlias) else tp
 
   def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation =

--- a/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala
@@ -8,6 +8,7 @@ import Types._
 import Contexts.Context
 import Symbols._
 import Decorators._
+import Denotations.SingleDenotation
 import SymDenotations.SymDenotation
 import DenotTransformers._
 import TypeUtils._
@@ -17,7 +18,7 @@ object ElimOpaque {
 }
 
 /** Rewrites opaque type aliases to normal alias types */
-class ElimOpaque extends MiniPhase with SymTransformer {
+class ElimOpaque extends MiniPhase with DenotTransformer {
 
   override def phaseName: String = ElimOpaque.name
 
@@ -27,14 +28,15 @@ class ElimOpaque extends MiniPhase with SymTransformer {
   // base types of opaque aliases change
   override def changesBaseTypes = true
 
-  override def transformNonSymInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type =
-    if (sym.isOpaqueHelper) TypeAlias(tp.extractOpaqueAlias) else tp
-
-  def transformSym(sym: SymDenotation)(implicit ctx: Context): SymDenotation =
-    if (sym.isOpaqueHelper) {
-      sym.copySymDenotation(
-        info = TypeAlias(sym.opaqueAlias),
-        initFlags = sym.flags &~ (Opaque | Deferred))
+  def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation =
+    if (ref.symbol.isOpaqueHelper)
+      ref match {
+        case sym: SymDenotation =>
+          sym.copySymDenotation(
+            info = TypeAlias(sym.opaqueAlias),
+            initFlags = sym.flags &~ (Opaque | Deferred))
+        case _ =>
+          ref.derivedSingleDenotation(ref.symbol, TypeAlias(ref.info.extractOpaqueAlias))
     }
-    else sym
+    else ref
 }

--- a/compiler/src/dotty/tools/dotc/transform/Getters.scala
+++ b/compiler/src/dotty/tools/dotc/transform/Getters.scala
@@ -65,6 +65,10 @@ class Getters extends MiniPhase with SymTransformer {
         d.copySymDenotation(
           initFlags = d.flags | maybeStable | AccessorCreationFlags,
           info = ExprType(d.info))
+            // Note: This change will only affect the SymDenotation itself, not
+            // SingleDenotations referring to a getter. In this case it does not
+            // seem to be a problem since references to a getter don't care whether
+            // it's a `T` or a `=> T`
       }
       else d
 

--- a/compiler/src/dotty/tools/dotc/transform/TypeUtils.scala
+++ b/compiler/src/dotty/tools/dotc/transform/TypeUtils.scala
@@ -51,5 +51,17 @@ object TypeUtils {
     /** The `*:` equivalent of an instance of a Tuple class */
     def toNestedPairs(implicit ctx: Context): Type =
       TypeOps.nestedPairs(tupleElementTypes)
+
+    /** Extract opaque alias from TypeBounds type that combines it with the reference
+     *  to the opaque type itself
+     */
+    def extractOpaqueAlias(implicit ctx: Context): Type = self match {
+      case TypeBounds(lo, _) =>
+        def extractAlias(tp: Type): Type = tp match {
+          case OrType(alias, _) => alias
+          case self: HKTypeLambda => self.derivedLambdaType(resType = extractAlias(self.resType))
+        }
+        extractAlias(lo)
+    }
   }
 }

--- a/tests/pos/i6286.scala
+++ b/tests/pos/i6286.scala
@@ -1,0 +1,6 @@
+object A {
+  opaque type T33 = Int
+  object T33 {
+    val a = new Array[T33](3)
+  }
+}

--- a/tests/pos/i6287.scala
+++ b/tests/pos/i6287.scala
@@ -1,0 +1,13 @@
+object O{
+  def m() = {
+    opaque type T = Int
+    object T
+  }
+}
+object A {
+  {
+    opaque type T = Int
+    object T
+    println
+  }
+}

